### ISSUE TITLE
Adding justified text

### ISSUE THREAD

**Issue by dr2p** [JujuAdams/Scribble#124]
Scribble is great and would be even better if it could justify text. I realize this is tricky since I couldn't come up with a perfect solution. 

Here's my attempt to justify text by modifying the scale on spaces within the string so that all the empty space at the end of each line is used by increasing the scale on previous spaces within the line. While it justifies most lines correctly, it's not perfect and breaks when adding formatting commands. A better solution is required.

Example of left aligned text:
![Left aligned text 2](https://user-images.githubusercontent.com/29870204/92204154-183a1a00-ee40-11ea-916e-7df591319032.png)

Example of "justified" text:
![justified text 2](https://user-images.githubusercontent.com/29870204/92204162-1c663780-ee40-11ea-9606-b1df83c04953.png)

Below is the code for the "justify_scribble_cache()" function which modifies the string within the scribble_cache to justify text. Perhaps it will be of some use for adding this feature.

```
function justify_scribble_cache(){
	
	var _scribble_cache = argument[0];
	var _page_width = _scribble_cache[@ SCRIBBLE.WIDTH];
	
	var _string = _scribble_cache[@ SCRIBBLE.STRING];
	var _string_mod = _string;
	var _space_width = scribble_get_width(" ");
	var _idx_mod = 0; // index modifier for manipulating _string_mod
	
	// Parse each line on each page to determine how much to scale each space to justify text  
	var _page_num = _scribble_cache[@ SCRIBBLE.PAGES];
	
	for (var _page=0; _page < _page_num; _page += 1)
	{
		//var _line_num = _scribble_cache[@ SCRIBBLE.PAGES_ARRAY][_page][@ __SCRIBBLE_PAGE.LINES]; // Not always correct! Check Scribble array for real answer  
		var _line_num = array_length(_scribble_cache[@ SCRIBBLE.PAGES_ARRAY][_page][@ __SCRIBBLE_PAGE.LINES_ARRAY]);
		if (_scribble_cache[@ SCRIBBLE.PAGES_ARRAY][_page][@ __SCRIBBLE_PAGE.LINES_ARRAY][_line_num-1] == undefined)
		{
			_line_num -= 1; 
		}
		var _page_width = _scribble_cache[@ SCRIBBLE.PAGES_ARRAY][_page][@ __SCRIBBLE_PAGE.WIDTH];
			
		for (var _line=0; _line < _line_num; _line += 1) // Don't scale spaces in the last line
		{
			var _start_char_line = _scribble_cache[@ SCRIBBLE.PAGES_ARRAY][_page][@ __SCRIBBLE_PAGE.LINES_ARRAY][_line][@ __SCRIBBLE_LINE.START_CHAR];
			var _last_char_line = _scribble_cache[@ SCRIBBLE.PAGES_ARRAY][_page][@ __SCRIBBLE_PAGE.LINES_ARRAY][_line][@ __SCRIBBLE_LINE.LAST_CHAR];
			var _char_count_line = _last_char_line - _start_char_line + 1; // index starts at 0 in scribble so +1 for index 0
			var _line_string = string_copy(_string, _start_char_line + 1, _char_count_line); // index starts at 1 for string_copy()
			var _line_width = scribble_get_width(_line_string);
			
			// Code hack below needed for better results. Perhaps a bug in scribble_get_width() for lines with line width from array greater than page width
			var _line_width_array = _scribble_cache[@ SCRIBBLE.PAGES_ARRAY][_page][@ __SCRIBBLE_PAGE.LINES_ARRAY][_line][@ __SCRIBBLE_LINE.WIDTH];
			if (_line_width_array >= _page_width)
			{
				_line_width = _line_width_array;
			}
			
			// Determine the number of spaces in line, the empty line space, and new scale for spaces for justified text 
			var _space_num = string_count(" ", _line_string) - 1; 
			var _empty_line_space = _page_width - _line_width; 

			if (_empty_line_space <= 0) 
			{
				var _space_scale_line = string_format(0.95, 1, 12); // This shouldn't be necessary but it is for better results
			}
			else if (_empty_line_space > 0)
			{
				var _space_scale_line = string_format((1 + ((_empty_line_space / _space_width) / _space_num)), 1, 12);
			}			
			
			for (var _idx=_start_char_line + 1; _idx < _last_char_line; _idx += 1) // First character shouldn't be a space
			{
				if (string_char_at(_string, _idx) == " ")
				{
					_string_mod = string_insert("[scale,]", _string_mod, _idx + _idx_mod);
					_string_mod = string_insert(_space_scale_line, _string_mod, _idx + _idx_mod + 7);
					_idx_mod = _idx_mod + 22; 
					_string_mod = string_insert("[/scale]", _string_mod, _idx + _idx_mod + 1);
					_idx_mod = _idx_mod + 8;
				}
			}
		}
	}

	return _string_mod;
}
```

I also added 
```
scribble_set_line_height(scribble_get_height("t"), scribble_get_height("t")); 
```
in the create event since increasing the scale on spaces will increase the height of text which affects line spacing.

**Comment by dr2p** [comment]
FYI- All lines of text in the example above were properly justified (including lines 4 and 22) by preventing text from going to multiple pages using the code below.
```
scribble_set_wrap(page_width, -1, false);
```

Something strange happens when text is being formatted into multiple pages. The first few characters in the first word of the 2nd page are not being displayed.

**Comment by JujuAdams** [comment]
Implemented for v7.2.0pre002

**Comment by DragoniteSpam** [comment]
okay, i think i might have some fun with this
